core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,415 @@
+            close0(null);
+            requested--;
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            if (requested <= 0) {
+                return;
+            }
+            subscription.request(1);
+        }
+
+        private void publishDownstream(T item) {
+            requireNonNull(item, "item");
+            if (closed) {
+                return;
+            }
+            downstream.onNext(item);
+            requested--;
+            if (!publishedAny) {
+                publishedAny = true;
+            }
+
+            publish();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (closed) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream(subscription);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            publishDownstream(item);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            close(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            setState(State.REQUIRE_BODY, State.REQUIRE_TAIL);
+            publish();
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (notifyCancellation) {
+                downstream.onError(cause);
+            }
+            downstream = NoopSubscriber.get();
+            completionFuture.completeExceptionally(cause);
+            release();
+        }
+
+        private void close(@Nullable Throwable cause) {
+            if (executor.inEventLoop()) {
+                close0(cause);
+            } else {
+                executor.execute(() -> close0(cause));
+            }
+        }
+
+        private void close0(@Nullable Throwable cause) {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            if (cause == null) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                downstream.onError(cause);
+                completionFuture.completeExceptionally(cause);
+            }
+            release();
+        }
+
+        private void release() {
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+                upstream.cancel();
```

Comment:
If the upstream publisher sent `onComplete()` or `onError()` signal, `upstream.cancel()` would be redundant.

Reply:
```java
private void close0 {
  if (closed) {
    return;
  }
  closed = true;
  ...
  release(); // release() is called only once, when closed == false
```
Hmm we call `release()` only once, when `closed == false`. isn't it?

```java
private void release() {
  if(closed) {
    return;
  }
  ..
}
```
Or I think we can check `closed` inside of release() 🤔

Reply:
I think the following call flow invokes `upstream.cancel()` even though the publisher sends `onComplete()` signal. If the publisher sends a complete signal I don't see we need to call `upstream.cancel()`.

`onComplete()` -> `publish()` -> `sendTail()` -> `sendComplete()` -> 
`close0(null)` (`closed` seems `false` when `close0()` is called -> `release()` -> `upstream.cancel()`.